### ISSUE TITLE
client sometimes doesn't return an error object for non-successful status codes

### ISSUE THREAD

**Issue by Janpot** [google/google-api-nodejs-client#467]
For example:

``` js
plus.people.get({ userId: ']' }, function (err, body, response) {
  console.log('err:')
  console.log(err)
  console.log('body:')
  console.log(body)
  console.log('statusCode:')
  console.log(response.statusCode)
});
```

This prints:

``` html
err:
null
body:
<HTML>
<HEAD>
<TITLE>Bad Request</TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" TEXT="#000000">
<H1>Bad Request</H1>
<H2>Error 400</H2>
</BODY>
</HTML>
statusCode:
400
```

Instead of returning an `err` with code `400`.

**Comment by jmdobry** [comment]
This works, note the use of `encodeURIComponent`:

```js
plus.people.get({
  userId: encodeURIComponent(']')
}, function (err, result) {
  if (err) return console.error(err.code);
  console.log(result)
});
```

When the parameters you're passing into the request contain non-alpha-numeric characters, you might need to encode them. On the other hand, perhaps this library should make sure they get properly encoded. This may be related to #476 and #499.

Closing as a duplicate, tracking this bug in those issues.

**Comment by Janpot** [comment]
@jmdobry I don't think I was clear on what was the issue here exactly, the bad input is only an example, I know how to encode it. The problem is that the function doesn't return with an error, `err` is just `null` in this case while it should be an `Error` object with a `code` property, like other errors. Can you reopen?

**Comment by jmdobry** [comment]
The scope of this issue as opened: `GET https://www.googleapis.com/plus/v1/people/]` returns 200 and some html instead of 400 and a useful error message. We can't change that the API returns 200 for this particular request (which is why you're not seeing an error object), but we can make this library more user friendly by catching bad input before making the request.

The bigger issue which is being tracked elsewhere: Improperly encoded path parameters can cause the request url to be invalid. See #499

The biggerer issue which is being tracked elsewhere: google-api-nodejs-client does not validate user input according to the discovery doc schema. See #528 

#499 and #528 encompass this issue, so we can safely mark this issue as a duplicate. Thanks for noting noting the problem!

**Comment by Janpot** [comment]
@jmdobry Yes I agree, path parameters need to be encoded, we agree on that.
But again, even though that would solve my problem here too, it is not exactly the scope of this issue, look at the following code:

```
plus.people.get({ userId: ']' }, function (err, body, res) {
  console.log(res.statusCode)
});
```

this outputs `null 400`. I hope we can agree that this is a contradiction and needs to be addressed on top of encoding path parameters.

For me it's mainly about robustness. My production app started crashing because it got strings where it expected an object. If the API would've handled 400s correctly, the encoding bug wouldn't have affected me. And any future similar bugs won't either.

**Comment by jmdobry** [comment]
In that case I'll re-open this issue and rename it to match what it's actually about.

**Comment by Janpot** [comment]
Thanks. Sorry I wasn't clear enough in the beginning. When I opened the issue I didn't realise this was actually two issues. The one bug exposed the other...

**Comment by Janpot** [comment]
I believe this is the problem:
https://github.com/google/google-auth-library-nodejs/blob/1f848ac4f136c2d2f2d0eddb331329c19cfa7dd5/lib/transporters.js#L95

I think 
```js
} else if (res.statusCode >= 500) {
``` 
is supposed to be 
```js
} else if (res.statusCode !== 200) {
```

**Comment by JustinBeckwith** [comment]
With the move over to axios in #891 , and with the encodeURIComponent fix in #275,  both issues should be solved now  🎉
